File: reference-implementations/java/src/main/java/org/tomlschema/ValidationResult.java
package org.tomlschema;

import java.util.List;
import java.util.stream.Stream;

/**
 * The errors and warnings produced by a validation attempt. SPEC.md requires that
 * implementations "provide separate access to errors and warnings"; both are the same
 * {@link ValidationDiagnostic} record, partitioned here by severity.
 *
 * @param errors the error-severity diagnostics
 * @param warnings the warning-severity diagnostics
 */
public record ValidationResult(List<ValidationDiagnostic> errors, List<ValidationDiagnostic> warnings) {
    /**
     * Creates a result with errors and no warnings.
     *
     * @param errors the validation errors
     */
    public ValidationResult(List<ValidationDiagnostic> errors) {
        this(errors, List.of());
    }

    public ValidationResult {
        errors = List.copyOf(errors);
        warnings = List.copyOf(warnings);
    }

    /**
     * Reports whether validation produced no errors.
     *
     * @return {@code true} when validation succeeded
     */
    public boolean isValid() {
        return errors.isEmpty();
    }

    /**
     * Returns all errors and warnings in diagnostic order.
     *
     * @return the combined diagnostics
     */
    public List<ValidationDiagnostic> diagnostics() {
        return Stream.concat(errors.stream(), warnings.stream()).toList();
    }
}

File: reference-implementations/dotnet/TomlSchema/ValidationResult.cs
namespace TomlSchema;

/// <summary>
/// The errors and warnings produced by a validation attempt. SPEC.md requires that
/// implementations "provide separate access to errors and warnings"; both are the same
/// <see cref="ValidationDiagnostic"/> record, partitioned here by severity.
/// </summary>
public record ValidationResult(
    IReadOnlyList<ValidationDiagnostic> Errors,
    IReadOnlyList<ValidationDiagnostic> Warnings)
{
    /// <summary>Creates a result with errors and no warnings.</summary>
    public ValidationResult(IReadOnlyList<ValidationDiagnostic> errors)
        : this(errors, Array.Empty<ValidationDiagnostic>()) { }

    /// <summary>Gets whether validation produced no errors.</summary>
    public bool IsValid => Errors.Count == 0;

    /// <summary>Gets errors and warnings in diagnostic order.</summary>
    public IEnumerable<ValidationDiagnostic> Diagnostics => Errors.Concat(Warnings);
}
